### PR TITLE
[65CE02] Chained INW/DEW for small constant offsets

### PR DESCRIPTION
**Pull request by mlund**
## Summary

- Emit chained `G_INC/G_DEC` for small constant pointer and integer additions on 65CE02+
- Post-RA, each `G_INC/G_DEC` becomes INW/DEW when register allocator places bytes in adjacent ZP pairs
- Asymmetric thresholds: ptr ±4, integer increment +2, integer decrement -1

Motivating pattern: strength-reduced `ptr[i*2]` loop access (e.g. `fire_fcm.cc`) becomes `ptr += 2` per iteration → `inw zp; inw zp` (4 bytes) instead of `lda/clc/adc #2/sta/lda/adc #0/sta` (14 bytes).

## Before/after (45GS02 vs 6502, typical cycles, INW/DEW = 7 cyc)

| Operation | 6502 bytes | 45GS02 bytes | 6502 cyc | 45GS02 cyc |
|-----------|-----------|-------------|---------|-----------|
| i16 += 2  | 10        | 9  (-1)     | 20      | 16  (-4)  |
| ptr += 2  | 14        | 5  (-9)     | 24      | 20  (-4)  |
| ptr += 3  | 14        | 7  (-7)     | 24      | 27  (+3)  |
| ptr += 4  | 14        | 9  (-5)     | 24      | 34 (+10)  |
| ptr -= 2  | 14        | 5  (-9)     | 24      | 20  (-4)  |
| ptr -= 4  | 14        | 9  (-5)     | 24      | 34 (+10)  |
| i32 += 2  | 22        | 17  (-5)    | 36      | 16 (-20)  |

## Real-world: `fire_fcm.cc` flame effect in full color mode on MEGA65

Chained `inw zp; inw zp` confirmed at two hot sites (stride-2 pointer access):
- `setup_screen`: screen pointer += 2 (1000 iterations)
- `convert_and_dma`: fire row pointer += 2 (1000 iterations/frame)

Per-site savings: 10 bytes, 5 instructions, 10 cycles per iteration.
Per frame: ~20,000 fewer cycles across both sites

Screenshot 2026-02-18 at 17 54 56

## Test plan

- [x] 77/77 MOS CodeGen tests pass
- [x] No change to 6502 codegen (thresholds gated on `has65CE02()`)
- [x] Edge cases: sub, add 0, i64, i8, boundary values
- [x] Verified chained INW in real-world `fire_fcm.cc` hot loops
- [x] **Update Aug. 2026**: Using a [patched SDK simulator with 65CE02 opcodes](https://github.com/mlund/llvm-mos-sdk/tree/sim-65ce02), this PR passes the complete [llvm-test-suite](https://github.com/llvm-mos/llvm-test-suite) and all tests in [mega65-freezer](https://github.com/mlund/mega65-freezer) a ~10k line C23 port of the original cc65 code.

Code changes were aided by Claude Opus 4.6 LLM and humanly supervised and reviewed
by @mlund in accordance with LLVM guidelines. I have tried to the best my knowledge to reduce any side effects and make the patch as confined as possible. The LLM discovered several flaws which are now caught by the embedded tests.

### CODE REVIEW

**Comment by mysterymath** [comment]
Is there a concern that is causing you to leave these two PRs drafted?

File: llvm/lib/Target/MOS/MOSLegalizerInfo.cpp [hunk 1]
```diff
@@ -1465,9 +1488,24 @@ bool MOSLegalizerInfo::legalizePtrAdd(LegalizerHelper &Helper,
     return true;
   }
 
-  if (ConstOffset && ConstOffset->Value.abs().isOne()) {
-    Builder.buildInstr(ConstOffset->Value.isOne() ? MOS::G_INC : MOS::G_DEC,
-                       {Result}, {Base});
+  // 65CE02+ pointers always live in ZP (Imag16), so chained INW/DEW (2 bytes,
```
Comment (mysterymath) [inline llvm/lib/Target/MOS/MOSLegalizerInfo.cpp]: Similar to the above: how bad does this get if the register allocator can't make this work?

File: llvm/lib/Target/MOS/MOSLegalizerInfo.cpp [hunk 2]
```diff
@@ -668,29 +668,52 @@ bool MOSLegalizerInfo::legalizeAddSub(LegalizerHelper &Helper,
...
   if (MI.getOpcode() == MOS::G_SUB)
     Amt = -Amt;
 
+  // 65CE02+ chained INC saves 1 byte and 4 cycles vs ADC for i16 += 2 when
+  // the value lives in A:X. Decrement threshold stays at 1 because chained
+  // DEC needs cmp #255 for borrow detection, adding 3 bytes per iteration.
+  const MOSSubtarget &STI = Builder.getMF().getSubtarget<MOSSubtarget>();
+  unsigned MaxIncDec = (STI.has65CE02() && Amt > 0) ? 2 : 1;
+
+  unsigned AbsAmt = static_cast<unsigned>(std::abs(Amt));
```
[the remark this answers is not in the document]
Comment (mlund) [reply llvm/lib/Target/MOS/MOSLegalizerInfo.cpp]: fixed together with the overflow above it that copolot missed

File: llvm/lib/Target/MOS/MOSLegalizerInfo.cpp (hunk 1 above)
Comment (mlund) [reply llvm/lib/Target/MOS/MOSLegalizerInfo.cpp]: Not for pointers which are always a zero-page pair, neighbors by construction and each step become one INW/DEW. I did benchmarking on mega65 hardware using the SDK examples and discovered several regressions compared to the 6502 path: One INW/DEW beats a 16-bit add. Two is a tie on speed and smaller. Three or four are slower than the add they replace and the effect is significant in a tight inner loop. In the original PR, the limit was four, which is now lowered to two. This keeps the pointer gain, but drops the regression. Here's the tables showing clear wins for loops due to ptr can be resident as a ZP pair:

##### Isolated (single operation, mos45gs02):

| operation | bytes main | bytes 551 | Δ | cycles main | cycles 551 | Δ |
|---|---|---|---|---|---|---|
| `i8 ±1` | 2 | 2 | · | 8 | 8 | · |
| `i8 ±2` | 4 | 4 | · | 10 | 10 | · |
| `i16 +1` | 5 | 5 | · | 11 | 11 | · |
| `i16 −1` | 7 | 7 | · | 13 | 13 | · |
| `i16 ±2` | 10 | 10 | · | 20 | 20 | · |
| `i32 +1` | 9 | 9 | · | 11 | 11 | · |
| `i32 −1` | 13 | 13 | · | 13 | 13 | · |
| `i32 ±2` | 22 | 22 | · | 36 | 36 | · |
| `i64 +1` | 17 | 17 | · | 11 | 11 | · |
| `i64 −1` | 45 | 45 | · | 13 | 13 | · |
| `i64 ±2` | 46 | 46 | · | 68 | 68 | · |
| `ptr ±1` | 3 | 3 | · | 13 | 13 | · |
| **`ptr ±2`** | **14** | **5** | **−9** | **24** | **20** | **−4** |
| `ptr ±3` | 14 | 14 | · | 24 | 24 | · |
| `ptr ±4` | 14 | 14 | · | 24 | 24 | · |

##### Loop (per iteration, mos45gs02):

| loop body | bytes main | bytes 551 | Δ | cycles main | cycles 551 | Δ |
|---|---|---|---|---|---|---|
| `i8 ±1` | 22 / 21 | 22 / 21 | · | 27 / 25 | 27 / 25 | · |
| `i8 ±2` | 23 | 23 | · | 27 | 27 | · |
| `i16 +1` | 30 | 30 | · | 39 | 39 | · |
| `i16 −1` | 55 | 55 | · | 76 | 76 | · |
| `i16 ±2` | 47 / 49 | 47 / 49 | · | 64 / 67 | 64 / 67 | · |
| `i32 +1` | 63 | 63 | · | 85 | 85 | · |
| `i32 −1` | 83 | 83 | · | 98 | 98 | · |
| `i32 ±2` | 71 / 77 | 71 / 77 | · | 108 / 112 | 108 / 112 | · |
| `i64 +1` | 81 | 81 | · | 115 | 115 | · |
| `i64 −1` | 137 | 137 | · | 155 | 155 | · |
| `i64 ±2` | 115 / 121 | 115 / 121 | · | 184 / 188 | 184 / 188 | · |
| `ptr +1` | 41 | 41 | · | 57 | 57 | · |
| `ptr −1` | 24 | 24 | · | 32 | 32 | · |
| **`ptr +2`** | **43** | **24** | **−19** | **55** | **37** | **−18** |
| **`ptr −2`** | **43** | **26** | **−17** | **55** | **39** | **−16** |
| `ptr ±3` | 43 | 43 | · | 55 | 55 | · |
| `ptr ±4` | 43 | 43 | · | 55 | 55 | · |

(INW/DEW = 7 cycles - MEGA65 user guide)
